packages/ERTP/src/keeper.js

```diff
@@ -0,0 +1,201 @@
+import harden from '@agoric/harden';
+
+import makeStore from '@agoric/weak-store';
+import { insist } from '@agoric/insist';
+
+const makeKeeper = (brand, unitOps) => {
+  const paymentLedger = makeStore('payment');
+  const purseLedger = makeStore('purse');
+
+  const makePayment = memo =>
+    harden({
+      getAllegedBrand: () => brand,
+      getMemo: () => memo,
+    });
+
+  // Methods like deposit() have an optional second parameter `units`
+  // which, if present, is supposed to be equal to the balance of the
+  // payment. This helper function does that check.
+  const insistExactly = (paymentBalance, units) => {
+    if (units !== undefined) {
+      insist(
+        unitOps.equals(units, paymentBalance),
+      )`payment balance ${paymentBalance} must equal units ${units}`;
+    }
+  };
+
+  const makePurse = memo => {
+    const purse = harden({
+      deposit: (paymentP, units = undefined) =>
+        Promise.resolve(paymentP).then(fromPayment => {
+          const fromPaymentBalance = paymentLedger.get(fromPayment);
+          insistExactly(fromPaymentBalance, units);
+          const purseBalance = purse.getBalance();
+          const newPurseBalance = unitOps.with(
+            fromPaymentBalance,
+            purseBalance,
+          );
+          // Commit point
+          // eslint-disable-next-line no-use-before-define
+          reallocate({
+            payments: [fromPayment],
+            purses: [purse],
+            newPurseBalances: [newPurseBalance],
+          });
+          return newPurseBalance;
+        }),
+      withdraw: units => {
+        const purseBalance = purse.getBalance();
+        const newPurseBalance = unitOps.without(purseBalance, units);
+        // Commit point
+        // eslint-disable-next-line no-use-before-define
+        const [payment] = reallocate({
+          purses: [purse],
+          newPurseBalances: [newPurseBalance],
+          newPaymentBalances: [units],
+        });
+        return payment;
+      },
+      getBalance: () => purseLedger.get(purse),
+      getAllegedBrand: () => brand,
+      getMemo: () => memo,
+    });
+    return purse;
+  };
+
+  // Units in circulation remain constant with reallocate
+  const reallocate = ({
+    purses = [],
+    payments = [],
+    newPurseBalances = [],
+    newPaymentBalances = [],
+    newPaymentMemos = [],
+  }) => {
+    const totalPursesBalance = purses
+      .map(purseLedger.get)
+      .reduce(unitOps.with, unitOps.empty());
+    const totalPaymentsBalance = payments
+      .map(paymentLedger.get)
+      .reduce(unitOps.with, unitOps.empty());
+    const total = unitOps.with(totalPursesBalance, totalPaymentsBalance);
+
+    const newPaymentsTotal = newPaymentBalances.reduce(
```

Comment:
Good idea. I think we should reserve the uppercase for constant magic numbers, since we're just renaming here, but what if we did:

```
    const add = unitOps.with;
    const empty = unitOps.empty();
    const totalPursesB = purses.map(purseLedger.get).reduce(add, empty);
    const totalPaymentsB = payments.map(paymentLedger.get).reduce(add, empty);
    const total = unitOps.with(totalPursesB, totalPaymentsB);

    const newPaymentsTotal = newPaymentBalances.reduce(add, empty);
    const newPursesTotal = newPurseBalances.reduce(add, empty);
    const newTotal = unitOps.with(newPaymentsTotal, newPursesTotal);
```

Reply:
That's a significant improvement, and the kind of thing I was looking for. I think it makes those parallel calls much easier to proofread.

I'm used to upper case for all manifest constants, so unitOps.with and unitOps.empty() would qualify. But I don't mind if the convention is different here.

Reply:
```js
const { with: add, empty } = unitOps;
```
I wish I could suggest
```js
const { with, empty } = unitOps;
```
except `with` is a keyword :(